ObjectiveGit/GTRepository.m

```diff
@@ -919,4 +920,76 @@ - (BOOL)cleanupStateWithError:(NSError * _Nullable __autoreleasing *)error {
 	return YES;
 }
 
+#pragma mark Notes
+
+- (GTNote *)createNote:(NSString *)note target:(GTObject *)theTarget ref:(NSString *)ref author:(GTSignature *)author committer:(GTSignature *)committer overwriteIfExists:(BOOL)overwrite error:(NSError **)error {
+	git_oid oid;
+	
+	int gitError = git_note_create(&oid, self.git_repository, ref.UTF8String, author.git_signature, committer.git_signature, theTarget.OID.git_oid, [note UTF8String], overwrite ? 1 : 0);
+	if (gitError != GIT_OK) {
+		if (error != NULL) *error = [NSError git_errorFor:gitError description:@"Failed to create a note in repository"];
+		
+		return nil;
+	}
+	
+	return [[GTNote alloc] initWithTargetOID:theTarget.OID repository:self ref:ref];
+}
+
+- (BOOL)removeNoteFromObject:(GTObject *)parentObject ref:(NSString *)ref author:(GTSignature *)author committer:(GTSignature *)committer error:(NSError **)error {
+	int gitError = git_note_remove(self.git_repository, ref.UTF8String, author.git_signature, committer.git_signature, parentObject.OID.git_oid);
+	if (gitError != GIT_OK) {
+		if(error != NULL) *error = [NSError git_errorFor:gitError description:@"Failed to delete note from %@", parentObject];
+		return NO;
+	}
+	
+	return YES;
+}
+
+- (BOOL)enumerateNotesWithError:(NSError **)error ref:(NSString *)ref usingBlock:(void (^)(GTNote *note, GTObject *object, BOOL *stop))block
+{
+	git_note_iterator* iter = NULL;
+	
+	int gitError = git_note_iterator_new(&iter, self.git_repository, ref.UTF8String);
+	
+	if (gitError != GIT_OK)
+	{
+		if(error != NULL) *error = [NSError git_errorFor:gitError description:@"Failed to enumerate notes"];
+		return NO;
+	}
+	
+	git_oid note_id;
+	git_oid object_id;
+	BOOL success = YES;
+	
+	while (git_note_next(&note_id, &object_id, iter) == GIT_OK) {
+		NSError* lookupErr = nil;
+		
+		GTNote* note = 	[[GTNote alloc] initWithTargetGitOID:&object_id repository:self.git_repository ref:ref.UTF8String];
+		if (note == nil) {
+			if (error != NULL)
+				*error = [NSError git_errorFor:GIT_ENOTFOUND description:@"Cannot create note"];
+			success = NO;
+			break;
+		}
+		
+		GTObject* obj = [self lookUpObjectByGitOid:&object_id error:&lookupErr];
+		if (obj == nil && lookupErr != nil) {
+			if (error != NULL)
+				*error = lookupErr;
+			success = NO;
```

Comment:
This is more debatable, but (even though I think there are other places we use that idiom), this prevents us from progressing through failures. Arguably, the "Cocoa way" would be to pass the `NSError *` to the block, but since it's the first time I notice it, I'll let someone else chime in...